### PR TITLE
feat: Ban/Unban by node id in api/client

### LINKED ISSUE #2411: Ban/ Unban by node id in api/client
There is already a Ban(nodeid) network command (and associated method)

* [ ] add Unban(NodeId) network command and associated method (see Unban(Vec<IpAddr>))
* [ ] add node_ban_by_id api endpoint and associated cli command
* [ ] add node_unban_by_id api endpoint and associated cli command 
* [ ] open up and monitor the api wiki update related issue

For the client command enum, watch where you add it, it matters for the help display

### PR DESCRIPTION
**Pull request by aoudiamoncef**
resolves #2411

### CODE REVIEW

**Comment by aoudiamoncef** [comment]
Hi @AurelienFT, 

We have to take some decisions on API design before:

```rust
     /// Keep it ? or not
    /// Unban a peer by his node id
    UnbanId(NodeId),


    /// Unban a list of peer by their node id
    UnbanIds(Vec<NodeId>) or UnbanId(Vec<NodeId>) or Unban(Vec<NodeId>) and change Unban to UnbanIp(Vec<IpAddress>) ?
```

**Comment by AurelienFT** [comment]
> Hi @AurelienFT,
> 
> We have to take some decisions on API design before:
> 
> ```rust
>      /// Keep it ? or not
>     /// Unban a peer by his node id
>     UnbanId(NodeId),
> 
> 
>     /// Unban a list of peer by their node id
>     UnbanIds(Vec<NodeId>) or UnbanId(Vec<NodeId>) or Unban(Vec<NodeId>) and change Unban to UnbanIp(Vec<IpAddress>) ?
> ```

### For the Network message part : 

I think the current naming is not consistent. It should be `BanIps` and `UnbanIps` and we should have `BanNodeIds` and `UnbanNodeIds` to match the new endpoint of the JSON RPC.

### For the JSON RPC API part : 
I think it should match what has been done to ban and unban by ip_addr. It's vecs for the moment so let's keep vecs and make : 
`Vec<NodeId>` as parameters of the API endpoint. .
For the name of the method of the JSON RPC we call them `ban_by_node_id` and `unban_by_node_id` or `ban_by_nodeid` and `unban_by_nodeid` there is no current consensus on that.

**Comment by aoudiamoncef** [comment]
IMHO, we are  in `massa_node` crate, we could remove node prefix for `massa_api`. 

**For the JSON RPC API part :**

- `ban_by_id`
- `unban_by_id`
- `ban_by_ip`
- `unban_by_ip`

**For the Network message part :**

- BanNodeIds 
- UnbanNodeIds
- BanIps
- UnbanIps

**Comment by AurelienFT** [comment]
> IMHO, we should have a consensus about function naming, especially public ones.
> 
> **For the Massa client:**
> 
> * `node_ban_by_id`
> * `node_unban_by_id`
> * `node_ban_by_ip`
> * `node_unban_by_ip`
> 
> **For the Massa API:**
> 
> * `node_ban_by_id`
> * `node_unban_by_id`
> * `node_ban_by_ip`
> * `node_unban_by_ip`
> 
> **For the NetworkCommand part :** IMHO, it's better to add `Node` prefix to easily distinguish commands`
> 
> ```rust
>     NodeBanById(NodeId),
>     NodeBanByIds(Vec<NodeId>),
>     NodeBanByIps(Vec<IpAddr>),
>     NodeUnbanById(NodeId),
>     NodeUnbanByIds(Vec<NodeId>),
>     NodeUnbanByIps(Vec<IpAddr>)
> ```

I agree with that. cc @damip

**Comment by damip** [comment]
Sorry for the delay, I approve of the naming.

**Comment by aoudiamoncef** [comment]
I'll do the requested changes ASAP.

**Comment by AurelienFT** [comment]
> We should keep ` NodeBanByIds(Vec<NodeId>)` or `NodeUnbanById(Vec<NodeId>)` ?
> 
> I'll do the requested changes ASAP.

Keep the batched one so with plural : `NodeBanByIds(Vec<NodeId>)` and `NodeUnbanByIds(Vec<NodeId>)`

File: massa.iml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (AurelienFT) [inline massa.iml:1]: What is this file ?

File: massa-protocol-exports/src/tests/tools.rs [hunk 2]
```diff
@@ -358,12 +346,12 @@ pub async fn assert_banned_nodes(
...
             msg = network_controller
                    .wait_command(1000.into(), |cmd| match cmd {
-                       NetworkCommand::Ban(node) => Some(node),
+                       NetworkCommand::NodeBanByIds(node) => Some(node),
                        _ => None,
                    })
              =>  {
                  let banned_node = msg.expect("Nodes not banned before timeout.");
-                 nodes.drain_filter(|id| *id == banned_node);
+                 nodes.drain_filter(|id| *id == banned_node[0]);
```
Comment (AurelienFT) [inline massa-protocol-exports/src/tests/tools.rs:354]: Just checking only the first element and not checking the whole list ?

File: massa-network-worker/src/network_cmd_impl.rs [hunk 3]
```diff
@@ -426,3 +433,37 @@ pub async fn on_ask_for_operations_cmd(
...
+                    ids.insert(*target_conn_id);
+                }
+            }
+        }
+    }
+
+    Ok(ids)
+}
+
+fn get_ips(worker: &mut NetworkWorker, node: &NodeId) -> Result<Vec<IpAddr>, NetworkError> {
```
Comment (AurelienFT) [inline massa-network-worker/src/network_cmd_impl.rs:456]: I don't think it's possible for a node to have multiple ips. I think this should be an error case.

**Review by AurelienFT — COMMENTED** [review]
I'm ok with the change. Just need to test it in local to approve. Thank you for your contribution !

File: massa.iml (hunk 1 above)
Comment (aoudiamoncef) [reply massa.iml:1]: Sorry, it's shouldn't be included, I'm using Intellij idea as an IDE.

File: massa-protocol-exports/src/tests/tools.rs (hunk 2 above)
Comment (aoudiamoncef) [reply massa-protocol-exports/src/tests/tools.rs:354]: Don't hesitate to improve it. I got an issue with this one, there is only one item left, and 4 elements right, so the assertion failed.

Comment (aoudiamoncef) [reply massa-protocol-exports/src/tests/tools.rs]: I was a little bit confused, time based tests are un predictable , especially when debug mode is ON. I'lL try to fix it

**Comment by AurelienFT** [comment]
Tested works well. Nice job. Thank you for your contribution ❤️

**Comment by AurelienFT** [comment]
bors merge

**Review by AurelienFT — APPROVED** [review]

**Comment by AurelienFT** [comment]
bors merge

**Pull request merged**